### PR TITLE
ENH: determine wheel tags by Python interpreter introspection & tests and CI improvements

### PR DESCRIPTION
**Pull request by dnicolodi**
Looking at the wheel tags generation code in `packaging` and in `wheel` it looks like it is full of cruft to support old interpreter versions. I tried to see how much effort it takes to have the most minimal compliant implementation. This for now only tests that my implementation returns the same values as the `packaging`-based one. Pushing it here just to have the tests run in CI.

### CODE REVIEW

**Comment by dnicolodi** [comment]
The code turns out nice and simple. The most complex part is a workaround for some macOS SDK bug. It would be nice if more of this string fiddling would be implemented in the standard library, but it does not seem as fragile as I was fearing. @FFY00 It would be nice if `sysconfig` on Windows would expose `SOABI` too.

I went ahead and replaced the use of `packaging` code with this implementation for the generation of the tags. Please have a look. @eli-schwartz 

The tests failures on macOS are due to the fact that a version of `packaging` with the macOS SDK workaround has not been released yet. I'm not sure what the best solution is here.

File: tests/conftest.py
Comment (rgommers) [inline tests/conftest.py:19]: Since `packaging` is no longer used, I assume the `manylinux` part of this is also no longer needed?

The change seems right to me; neither `meson-python` nor `packaging` can know whether or not the wheel being produced is manylinux compliant, and there should be no attempt to deal with that at all (that's for `auditwheel`).

File: tests/test_tags.py
Comment (rgommers) [inline tests/test_tags.py:49]: These should be kept, right? We had a lot of trouble with wheel tags, so I'd like to make sure there are no regressions - that would make packages immediately uninstallable from source.

Comment (dnicolodi) [reply tests/test_tags.py:49]: This was testing the conversion from a extension module filename suffix to a wheel tag. This approach cannot work as the filename suffix does not carry enough information (and strictly speaking it is not mandatory for it to carry any). This patch set moves away from that approach. Thus this test is not necessary anymore and the code that is being tested does not exist anymore.

File: tests/conftest.py
Comment (dnicolodi) [reply tests/conftest.py:19]: This is only used in the tests to compare what `meson-python` generates with what `packaging` generates. `packaging` generates the very specific `manylinux` platform tag while `meson-python` does not. Thus we need to adapt the former version of the tag to the latter. However, while I was writing this I realized that this does not handle `muslinux` tags, thus the tests are broken on platforms using that libc. Fixing it.

Comment (rgommers) [reply tests/conftest.py:19]: Ah okay, got it. So we use `packaging` in the test suite to ensure that there's no silent mismatches that result in wheels we produce being uninstallable - that seems quite useful.

File: .github/workflows/tests.yml
Comment (DWesl) [inline .github/workflows/tests.yml]: The nox configuration also installs `pytest-github-actions-annotate-failures`; do you wish to install that as well?

File: .github/workflows/tests.yml
Comment (DWesl) [inline .github/workflows/tests.yml:122]: It looks like the nox configuration would also install `pytest-github-actions-annotate-failures`; do you want to install that as well?

**Review by DWesl — COMMENTED** [review]
The Cygwin CI configuration looks solid to me; I especially like using `/usr/bin/env` to make pytest's description of the test environment shorter.

File: .github/workflows/tests.yml
Comment (dnicolodi) [reply .github/workflows/tests.yml:84]: I put the Python version in the matrix mostly as a reminder of which Python version this job runs. I don't expect Cygwin to update the Python version they distribute anytime soon.

File: .github/workflows/tests.yml
Comment (dnicolodi) [reply .github/workflows/tests.yml:110]: In that case the wheels will be rebuilt. That's why weels are tagged with a platform specifier. I don't foresee any problem with this. The effort to put in place and maintain this is negligible and it cuts the job time in half, which is quite nice when debugging.

Comment (eli-schwartz) [reply .github/workflows/tests.yml:110]: As long as it dramatically speeds up the CI at least one in five CI runs, it sounds like a huge win to me?

I'm quite sure that there will be cache hits more than 20% of the time though. :D

File: .github/workflows/tests.yml
Comment (dnicolodi) [reply .github/workflows/tests.yml]: I dislike stuff that annotates the diffs on GitHub it makes reviewing them a pain. But I can reinstate this one if people like it.

**Comment by dnicolodi** [comment]
All green except mysterious failures on Homebrew with Python 3.10 and 3.11. I'm convinced these are test setup issues, but debugging the issues with the Cygwin job drained my divination power for the day.

File: .github/workflows/tests.yml
Comment (DWesl) [reply .github/workflows/tests.yml:84]: The other option would be to try testing on python 3.7 or 3.8, also currently available from Cygwin, and the reason the package is no longer named python3 and python3-pip.  I don't think that would turn up too many more bug, though.

**Comment by eli-schwartz** [comment]
```
  brew reinstall python@3.11 || true
```

This command fails, but is told to report the job as successful regardless of whether it actually succeeded. I'm not sure what the point of this is supposed to be, other than to make errors less obvious. :)

...

It's a github actions bug, github actions has inherently broken homebrew: https://github.com/Homebrew/discussions/discussions/3895

File: .github/workflows/tests.yml
Comment (DWesl) [reply .github/workflows/tests.yml:110]: My concern would be the new wheel don't get added to the cache; the next run continues to pull the old cache rather than getting a new version.  It seems the consensus is that the cache will be expired at least every couple of months, so this isn't really an issue.

File: .github/workflows/tests.yml
Comment (DWesl) [reply .github/workflows/tests.yml]: If it just lets you jump to the start of the detailed error reports in the CI logs, I can see the use for that; if it mostly clutters the diff that's less useful.

I just found it as a side effect of checking that the tests didn't use any of the `GITHUB_*` environment variables, since I remembered seeing one recently.

File: .github/workflows/tests.yml
Comment (dnicolodi) [reply .github/workflows/tests.yml:110]: The cache is keyed with the hash of the commit that triggered the CI job but retrieved with a less specific key in LRU fashion, thus this is not a concern. You can read the documentation for `actions/cache` to get the details.

Comment (DWesl) [reply .github/workflows/tests.yml:110]: That works.

**Comment by dnicolodi** [comment]
> ```
>   brew reinstall python@3.11 || true
> ```
> 
> This command fails, but is told to report the job as successful regardless of whether it actually succeeded. I'm not sure what the point of this is supposed to be, other than to make errors less obvious. :)

The `brew install` fails to link the package because there are already binaries from Python 2.7 in place of the destinations of the homebrew links for the unversioned binaries. The `brew link --overwrite` solves overwrites the existing binaries with the homebrew links. The other solution I played with was to `rm -f` the I'll make sure to stick a comment explaining what that does once I solve the remaining issues, if it is possible.

> It's a github actions bug, github actions has inherently broken homebrew:

Oh. Good to know. I'll read up about it.

**Comment by eli-schwartz** [comment]
> The `brew install` fails to link the package because there are already binaries from Python 2.7 in place of the destinations of the homebrew links for the unversioned binaries.

Actually the GitHub Actions specific 3.11, not 2.7, but otherwise correct.

File: .github/workflows/tests.yml
Comment (dnicolodi) [reply .github/workflows/tests.yml:84]: The Cygwin tests are interesting because of the peculiarities of the platform. I don't expect differences between Python versions. We already have many CI jobs and I'm more thinking or remove some elements of the test matrix rather than adding some more.

**Comment by dnicolodi** [comment]
The issues with the Homebrew Python are due to this https://github.com/pypa/pip/issues/11539 I still don't know if it is due to a bug in pip or something weird in how Homebrew Python is setup. It is strange that I don't seem to encounter this error in my local setup, though.

File: mesonpy/_tags.py
Comment (FFY00) [inline mesonpy/_tags.py:46]: The `SOABI` isn't documented anywhere, so could be removed in future Python versions. While I am okay merging this as an initial implementation, granted that all of our current tests pass, we should work on figuring out a better way to do this.

Please add an `# XXX: ...` comment documenting this.

File: mesonpy/_tags.py
Comment (FFY00) [inline mesonpy/_tags.py:49]: This should be a proper check, not an assertion. It may go haywire if the asserts are disabled and the implementation doesn't actually provide `SOABI`.

File: mesonpy/_tags.py
Comment (FFY00) [inline mesonpy/_tags.py:95]: nitpick: Make this a constant.

**Review by FFY00 — APPROVED** [review]
I have some concerns regarding the reliance on `SOABI`, as presented below, but overall this PR looks good.

File: mesonpy/_tags.py
Comment (FFY00) [inline mesonpy/_tags.py:54]: No special handling is needed for PyPy AFACT. Am I missing something?

File: mesonpy/_tags.py
Comment (dnicolodi) [reply mesonpy/_tags.py:38]: The idea is to rely on the value of `Py_DEBUG` if set (`sysconfig.get_config_var()` does not return Null) and fallback to the `hasattr` check otherwise. Thus open-coding this would be something like:
```
debug = _get_config_var('Py_DEBUG')
if debug is None:
    debug = hasattr(sys, 'gettotalrefcount')
```
Because we have two of these checks I decided to have the tiny helper function. Happy to open code it if you prefer.

File: mesonpy/_tags.py
Comment (dnicolodi) [reply mesonpy/_tags.py:46]: `packaging` relies on the same configuration variable. As we need to be bug-compatible with `packaging` I figure this is okay. AFAIK, none of the variables available with `sysconfig.get_config_var()` is documented. Should I add the same comment everywhere we use one?

File: mesonpy/_tags.py
Comment (dnicolodi) [reply mesonpy/_tags.py:95]: The only name for the constant I can think about is `_2_POWER_32`, which is maybe more cryptic than the literal value. Do you have a specific name that would be clearer than the value itself?

File: mesonpy/_tags.py
Comment (dnicolodi) [reply mesonpy/_tags.py:54]: I agree. However, the special handling is in `packaging` and I decided to keep it. I'm happy to remove it if we are positive that it is not needed.

File: mesonpy/_tags.py
Comment (DWesl) [reply mesonpy/_tags.py:95]: `sys.maxsize < 2 ** 32` would be equivalent to `sys.maxsize <= UINT32_MAX`; since this isn't Windows, that's likely the same as `sys.maxsize <= ULONG_MAX`.  I suspect this is being used as a delineation between 32-bit Python and 64-bit Python, so you could also choose a name based on that (`PYTHON_32BIT_SYS_MAXSIZE_CUTOFF`, maybe)

File: mesonpy/_tags.py
Comment (DWesl) [inline mesonpy/_tags.py:99]: It looks like `arch = 'i386'` will overwrite `arch = 'ppc'`; should that be before the conditional, or guarded by `else:`?

File: mesonpy/_tags.py
Comment (eli-schwartz) [reply mesonpy/_tags.py:95]: Maybe the suggestion was to use `if sys.maxsize < 4294967296` but I'm not sure whether that's all that much of an advantage...

Comment (FFY00) [reply mesonpy/_tags.py:95]: I mean, make `sys.maxsize <= 2**32` a constant, not `2**3`. Something like `_32_BIT_INTERPRETER`, just like `packaging` does.

File: mesonpy/_tags.py
Comment (FFY00) [reply mesonpy/_tags.py:46]: It does, but only for the non CPython tags. The normalized version of `SOABI` is used as the full ABI tag as a fallback, we can do the same.

IIRC we only use `SOABI` in one other place, and it has a fallback value.

File: mesonpy/_tags.py
Comment (FFY00) [reply mesonpy/_tags.py:54]: Where? The correspondent code AFAIK is

https://github.com/pypa/packaging/blob/606c71acce93d04e778cdbdea16231b36d6b870f/packaging/tags.py#L496-L499

Which only has special handling for CPython, as it's the only needed. For other implementations, the value falls back to the normalized version of `SOABI`.

Comment (FFY00) [reply mesonpy/_tags.py:54]: In fact, we won't generate the same tags as `packaging` if PyPy adds another section to `SOABI`, as this code is only taking the first two.

File: mesonpy/_tags.py
Comment (dnicolodi) [reply mesonpy/_tags.py:46]: We use other variables from `sysconfig.get_config_var()` all of them are undocumented. Aren't they?

File: mesonpy/_tags.py
Comment (dnicolodi) [reply mesonpy/_tags.py:95]: I'll do that but I really don't see the advantage.

File: mesonpy/_tags.py
Comment (henryiii) [reply mesonpy/_tags.py:54]: We should not be using SOABI. It is unreliable. It's better to use EXT_SUFFIX, and parse from there. Packaging is doing this in https://github.com/pypa/packaging/pull/607. Because packaging's usage of this has been wrong for so long, other implementiaonts might have broken `SOABI`'s too, just to make packing work. While `EXT_SUFFIX` is much more likely to be correct, as it's used in other places. Both variables are documented in [PEP 3149](https://peps.python.org/pep-3149/), and are supposed to be identical, minus the leading dot and extension. Also, Python itself recommends `EXT_SUFFIX` if you try to use `SO`.

Comment (dnicolodi) [reply mesonpy/_tags.py:54]: Sorry, I meant `wheel` https://github.com/pypa/wheel/blob/814c2efe8e40051039c5a6de6945e04ecdd162ee/src/wheel/bdist_wheel.py#L96-L99 I've been staring at all these implementations for too long. I agree that the special handling is weird. I'll remove it.

File: mesonpy/_tags.py
Comment (dnicolodi) [reply mesonpy/_tags.py:99]: Good catch! I guess I refactored this code around and I had this in a function with a `return` before.

File: mesonpy/_tags.py
Comment (henryiii) [reply mesonpy/_tags.py:46]: It is documented in [PEP 3149](https://peps.python.org/pep-3149/) and should be available on all CPythons, along with `EXT_SUFFIX`. As I mentioned on the other thread, `EXT_SUFFIX` is more reliable due to a long standing bug in packaging expecting a broken `SOABI` because PyPy has a broken SOABI. You can get a little backstory in https://foss.heptapod.net/pypy/pypy/-/issues/3816 or https://gitlab.kitware.com/cmake/cmake/-/issues/24121.

File: mesonpy/_tags.py
Comment (dnicolodi) [reply mesonpy/_tags.py:54]: @henryiii I think I don't understand. I cannot parse what you are trying to say with this sentence

> Because packaging's usage of this has been wrong for so long, other implementations might have broken `SOABI`'s too, just to make packing work (and therefor pip, etc).

Can you please try to rephrase it?

If PEP3149 documents both variables to carry the same information, if they would get out of sync, that would be a bug of the Python implementation. Wouldn't it? Can we find a Python implementation that has the wrong `SOABI` but the right `EXT_SUFFIX`?

Comment (henryiii) [reply mesonpy/_tags.py:54]: > Can we find a Python implementation that has the wrong `SOABI` but the right `EXT_SUFFIX`?

Absolutely: PyPy. PyPy reports the wrong SOABI, it's missing the platform information. If you use it, you make an extension that can't be loaded (since EXT_SUFFIX is used for loading). When Packaging was adapted to use SOABI, it special cases CPython, so the "general" code was based on PyPy's mistake, so it assumes SOABI is wrong. So other implementations might report wrong SOABI's because it makes packing work. It's being fixed in packaging right now (to use EXT_SUFFIX and assume it is right), but until the current Pip is old enough to not matter, PyPy is stuck reporting the wrong SOABI.

If you report the wrong `EXT_SUFFIX`, setuptools and probably other things break. If you report the _right_ `SOABI` but you are not CPython, you might not be loadable by Pip currently (is being fixed in that linked PR).

Comment (henryiii) [reply mesonpy/_tags.py:54]: According to the PEP, `assert sysconfig.get_config_var('EXT_SUFFIX').split(".", 1)[0].lstrip(".") == sysconfig.get_config_var('SOABI')`, but it doesn't for PyPy, so using the former is best. For just the ABI, yes, that would suffice.

What's wrong with CPython 3.7?

```console
$ docker run --rm -it quay.io/pypa/manylinux2014_x86_64
[root@f00ca105d347 /]# python3.7
Python 3.7.14 (default, Sep 12 2022, 15:16:12)
[GCC 10.2.1 20210130 (Red Hat 10.2.1-11)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sysconfg
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'sysconfg'
>>> import sysconfig
>>> sysconfig.get_config_var("EXT_SUFFIX")
'.cpython-37m-x86_64-linux-gnu.so'
>>> sysconfig.get_config_var("SOABI")
'cpython-37m-x86_64-linux-gnu'
>>>
[root@f00ca105d347 /]# pypy3.7
Python 3.7.13 (7e0ae751533460d5f89f3ac48ce366d8642d1db5, Mar 29 2022, 06:03:31)
[PyPy 7.3.9 with GCC 10.2.1 20210130 (Red Hat 10.2.1-11)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>>> import sysconfig
>>>> sysconfig.get_config_var("EXT_SUFFIX")
'.pypy37-pp73-x86_64-linux-gnu.so'
>>>> sysconfig.get_config_var("SOABI")
'pypy37-pp73'
```

Comment (eli-schwartz) [reply mesonpy/_tags.py:54]: My recollection that CPython started adding more complex EXT_SUFFIX instead of just ".so", as part of the migration from 2.7 to the entire 3.x series, and that change, like most 3.x changes, happened "sometime during 3.0, 3.1, 3.2, or 3.3, although most people aren't sure which of them because it took several releases before the general populace considered it to be a viable replacement for 2.7".

Comment (dnicolodi) [reply mesonpy/_tags.py:54]: > Was it an early patch release of 3.7? Or something else?

CPython 3.7 on Windows just reports the extension. See #190.

Comment (henryiii) [reply mesonpy/_tags.py:54]: PEP 3149 was accepted in 2010, so that's what I would have figured too, which is why I was asking about why a fallback was needed for 3.7.

Comment (dnicolodi) [reply mesonpy/_tags.py:54]: See https://docs.python.org/3.8/whatsnew/changelog.html#python-3-8-7-final and https://github.com/python/cpython/issues/84006

Comment (henryiii) [reply mesonpy/_tags.py:54]: Ah, you said `.so`. Okay, sounds like a fallback is needed for Windows and Python 3.7, then.

Comment (dnicolodi) [reply mesonpy/_tags.py:54]: Yes, my bad. Sorry.

Comment (dnicolodi) [reply mesonpy/_tags.py]: Nope. Mayne I misunderstood the proposal, but looking at `EXT_SUFFIX` and assuming is correct does not fly.

CPython
```
>>> sysconfig.get_config_var('EXT_SUFFIX')
'.cpython-310-darwin.so'
>>> next(packaging.tags.sys_tags()).abi
'cp310'
```

PyPy:
```
>>>> next(packaging.tags.sys_tags()).abi
'pypy37_pp73'
>>>> sysconfig.get_config_var('EXT_SUFFIX')
'.pypy37-pp73-darwin.so'
```

Pyston:
```
>>> print(sysconfig.get_config_var("EXT_SUFFIX"))
'.pyston-23-x86_64-linux-gnu.so'
>>> next(packaging.tags.sys_tags()).abi
'pyston_23_x86_64_linux_gnu'
```

Also, this is the beautiful code currently proposed for `packaging` in 
```python
def _generic_abi() -> List[str]:
    """Return the ABI tag based on EXT_SUFFIX"""
    # ext_suffix is something like this. We want to keep the parts which are
    # related to the ABI and remove the parts which are related to the
    # platform:
    # - linux:   '.cpython-310-x86_64-linux-gnu.so' => cp310
    # - mac:     '.cpython-310-darwin.so'           => cp310
    # - win:     '.cp310-win_amd64.pyd'             => cp310
    # - pypy:    '.pypy38-pp73-x86_64-linux-gnu.so' => pypy38-pp73
    # - graalpy: '.graalpy-38-native-x86_64-darwin.dylib'
    #                                               => graalpy-38-native

    ext_suffix = _get_config_var("EXT_SUFFIX", warn=True)
    assert ext_suffix[0] == "."
    _, soabi, ext = ext_suffix.split(".")
    if soabi.startswith("cpython"):
        abi = "cp" + soabi.split("-")[1]
    elif soabi.startswith("pypy"):
        abi = "-".join(soabi.split("-")[:2])
    elif soabi.startswith("graalpy"):
        abi = "-".join(soabi.split("-")[:3])
    elif soabi:
        abi = soabi
    else:
        return []
    return [_normalize_string(abi)]
```
which is many things but not generic. Lovely.

Comment (dnicolodi) [reply mesonpy/_tags.py]: I'm going to stick to `SOABI` just because it takes lees code. Is really a `sysconfig.get_interpreter_abi()` function too much to ask?

Comment (henryiii) [reply mesonpy/_tags.py]: What is the result of SOABI in those cases? It should be `cpython-310-darwin`, `pypy37-pp73-darwin` (but actually is `pypy37-pp73`), and `pyston-23-x86_64-linux-gnu`, correct?

Comment (henryiii) [reply mesonpy/_tags.py]: By "assuming it is correct", I mean that it's the correct extension suffix. `pypy37-pp73.so` for example is not a valid suffix, pypy can't load it. I didn't mean the wheel abi name is "better" from `EXT_SUFFIX`, I just meant `SOABI` is sometimes wrong, while `EXT_SUFFIX` is correct. You still need to transform it for wheel tag names in either case - and it's exactly the same transform unless SOABI is buggy. And if SOABI is buggy, the implementation will hopefully try to change it (as PyPy will in a year or two).

Comment (mattip) [reply mesonpy/_tags.py]: > I don't understand how going from special casing PyPy to special casing PyPy and GrallVM Python while still getting it wrong for Pyston is an improvement

The pyston developers are welcome to add to that code or provide a comment as to what needs to be added. In a perfect world, sysconfig would provide `get_python_tag()`, `get_abi_tag()` and `get_platform_tag()`, but it doesn't, so here we are. The goal of the PR you mentioned is to fix a possible issue when PyPy fixes its SOABI value at some time in the future. It was not intended to make the code generic (which is impossible as long as packaging needs to be backward-compatible) or to cover all implementations (that is up to other projects to contribute where needed).

Both pip and wheel now use `packaging/tags.py`. It is a shame this functionality is replicated in other places.

Comment (dnicolodi) [reply mesonpy/_tags.py]: > The pyston developers are welcome to add to that code or provide a comment as to what needs to be added.

How `packaging` generates tags for Pyston cannot be changed for compatibility: compatibility of existing wheel releases with systems upgraded to an eventual future `packaging` and compatibility of systems with an old packaging with wheels generated on a system with an eventual future `packaging`.

> Both pip and wheel now use `packaging/tags.py`. It is a shame this functionality is replicated in other places.

Should `packaging` start to provide `get_python_tag()`, `get_abi_tag()` and `get_platform_tag()` functions, we may consider it. Till then the API is inconvenient to use. The algorithm is set in stone for all existing Python implementations (see above) and new Python implementations do not yet pup up like mushrooms, so keeping it in sync is not much work at all.

Comment (dnicolodi) [reply mesonpy/_tags.py]: > What is the result of `SOABI` in those cases? It should be `cpython-310-darwin`, `pypy37-pp73-darwin` (but actually is `pypy37-pp73`), and `pyston-23-x86_64-linux-gnu`, correct?

Correct.

**Comment by dnicolodi** [comment]
@FFY00 I think this is now ready. After some serious debugging effort the CI should be happy now. Please take a look.

I would appreciate if we could cut a release after merging this, unless there are other important things pending. I would like to go back to release wheels for Python 3.7, which is the thing that originally prompted this work.

**Review by FFY00 — APPROVED** [review]
I think this looks good. Huge thanks for working on this @dnicolodi! :tada:

**Comment by henryiii** [comment]
> I would appreciate if we could cut a release after merging this

I'm happy to finish off #175 as soon as this goes if if you'd like that too - that's quite important for building on non-wheel-supported platforms.

**Comment by henryiii** [comment]
Great work, by the way! :)

**Comment by FFY00** [comment]
We have a couple things ready that are blocked by fixes on the PR, after we get this in we can merger all the blocked stuff and make a release.

**Comment by dnicolodi** [comment]
The system ninja work would be very nice to have indeed.

**Comment by henryiii** [comment]
This "passes" but the pytest -> test name change messes up the require checks. Personally, I prefer having a "pass" check that depends on all the other checks, and then only require that (once per workflow, unfortunately, but still much simpler than maintaining a bunch of checks, and if you change the checks all the old merged PRs don't change to orange dots).

**Comment by dnicolodi** [comment]
Yeah. Sorry, I didn't realize that the rename would have caused this issue. However, also the readthedocs job is stuck, so this is not the only thing preventing this from being merged automatically. Now that all the CI jobs are fixed, can't we simply require that all jobs succeed? I'm not sure why the status of old PRs should change changing the definition of the checks for future PRs. It seems a GitHub bug to me.

**Pull request merged**
